dask/typing.py

```diff
@@ -0,0 +1,335 @@
+from __future__ import annotations
+
+import abc
+from collections.abc import Callable, Hashable, Mapping, Sequence
+from typing import Any, Protocol, TypeVar, runtime_checkable
+
+try:
+    from IPython.display import DisplayObject
+except ImportError:
+    DisplayObject = Any
+
+
+CollType_co = TypeVar("CollType_co", bound="DaskCollection")
+CollType_contra = TypeVar("CollType_contra", bound="DaskCollection", covariant=True)
+PostComputeCallable = Callable
+
+
+class PostPersistCallable(Protocol[CollType_contra]):
+    """Protocol defining the signature of a __dask_postpersist__ method."""
+
+    def __call__(
+        self,
+        dsk: Mapping,
+        *args: Any,
```

Comment:
I think some of the pyright issues that I saw were due to this `*args`. The implementation for most of the collections don't actually accept arbitrary `*args`, making the function signatures incompatible.

I'm actually not sure why mypy is okay with it.

Reply:
Looking at this a bit more: it would be nice if we could type `*args` with a `ParamSpec`. However, I think `ParamSpec` is currently not expressive enough to handle keyword-only args: https://peps.python.org/pep-0612/#concatenating-keyword-parameters.

So I take the above back, I'm not sure we can do better than `Any` here at the moment.